lib/Db/TagMapper.php

```diff
@@ -0,0 +1,250 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Mail\Db;
+
+use function array_map;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Db\Entity;
+use OCP\AppFramework\Db\QBMapper;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+use OCP\IL10N;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @template-extends QBMapper<Tag>
+ */
+class TagMapper extends QBMapper {
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var IL10N */
+	private $l10n;
+
+	public function __construct(IDBConnection $db,
+								LoggerInterface $logger,
+								IL10N $l10n) {
+		parent::__construct($db, 'mail_tags');
+		$this->logger = $logger;
+		$this->l10n = $l10n;
+	}
+
+	/**
+	 * @param string $imapLabel
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTagByImapLabel(string $imapLabel, string $userId): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->eq('imap_label', $qb->createNamedParameter($imapLabel)),
+				$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+			);
+		return $this->findEntity($qb);
+	}
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTagForUser(int $id, string $userId): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+					$qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)),
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		return $this->findEntity($qb);
+	}
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
+	 * @return array
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getAllTagForUser(string $userId): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		return $this->findEntities($qb);
+	}
+
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTag(int $id): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)));
+		return $this->findEntity($qb);
+	}
+
+	/**
+	 * Tag a message in the DB
+	 *
+	 * To tag (flag) a message on IMAP, @see OCA\Mail\Service\MailManager::tagMessage
```

Comment:
```suggestion
	 * To tag (flag) a message on IMAP, @see \OCA\Mail\Service\MailManager::tagMessage
```

then IDEs like phpstorm will let us jump directly to the method